morph/cli.py

```diff
@@ -116,6 +131,109 @@ def cmd_count(args):
         print '%d\t%s' % (c, m.show().encode('utf-8'))
 
 
+def cmd_next(args):
+    notes_path = args.notes
+    prio_path = args.prio
+    mizer = MIZERS[args.mizer]
+
+    known = load_db('known')
+
+    candidates = defaultdict(list)
+    with codecs.open(notes_path, 'r', 'utf-8') as f:
+        for line in f.readlines():
+            note = line.strip()
+            text = note.split('\t', 1)[0]
+            unknowns = [m for m in mizer.getMorphemesFromExpr(text)
+                        if m not in known.db]
+            if len(unknowns) == 1:
+                candidates[unknowns[0].show()].append(note)
+
+    with codecs.open(prio_path, 'r', 'utf-8') as f:
+        for line in f.readlines():
+            freq, m = line.strip().split('\t', 1)
+            m_cite = m.split('\t', 1)[0]
+            for cand in candidates[m]:
+                print (u'%s\t%s\t%s' % (freq, m_cite, cand)).encode('utf-8')
+
+
+def cmd_grep(args):
+    pattern_string = args.pattern
+    files = args.files
+    max_count = args.max_count
+    mizer = MIZERS[args.mizer]
+
+    pattern = parse_morpheme(pattern_string.decode('utf-8'))
+
+    count = 0
+    for path in files:
+        with codecs.open(path, 'r', 'utf-8') as f:
+            for line in f:
+                ms = mizer.getMorphemesFromExpr(line)
+                if pattern in ms:
+                    sys.stdout.write(line.encode('utf-8'))
+                    count += 1
+                    if max_count is not None and count >= max_count:
+                        return
+
+
+def clear_locs(db, pred):
+    '''Remove from `db` locations that match `pred`, and morphemes left without locations.'''
+    empty_ms = []
+    for m, locs in db.db.iteritems():
+        locs.difference_update([loc for loc in locs if pred(loc)])
+        if not locs:
+            empty_ms.append(m)
+    for m in empty_ms:
+        del db.db[m]
+
+
+def cmd_sync_known(args):
+    filenames = args.input or ['known.txt']
+    should_merge = args.merge
+
+    external_db = load_db('external', allow_missing=True)
+    if not should_merge:
+        clear_locs(external_db, lambda loc: isinstance(loc, morph.morphemes.Nowhere))
+
+    db_dir = os.path.join(profile_path(), 'dbs')
+    loc = morph.morphemes.Nowhere(maturity=30)  # arbitrary but > maturity threshold
+    for filename in filenames:
+        with codecs.open(os.path.join(db_dir, filename), 'r', 'utf-8') as f:
+            for line in f:
+                line = line.strip()
+                if not line or line.startswith('#'):
```

Comment:
Ah, this should be documented, thanks. This actually is for the user -- the point is to be able to have comments in the file to organize it or help oneself understand later what's there and why.

For example, the way I've used this `sync-known` command is that I took the most frequent morphemes from an `mm count` on a corpus, minus (using `grep -v`) those that appeared in `mm dump known`, and in an editor started picking out the ones I actually knew and my Anki data just didn't know I knew, then passed that file to `mm sync-known`. Many of the top morphemes turn out, unsurprisingly, to be very short words with grammatical functions where there are several words/morphemes spelled similarly; because the complete description of the morpheme specifies the part of speech in some detail, I wanted to make sure the morpheme I was looking at was what I thought it was. Once I had, I wanted to leave notes for myself on what I'd found. So the first lines of the file look like this:
```
# as in である
ある	助動詞		アル

# the "parallel marker", ~"and"
と	助詞	並立助詞	ト

# the familiar ~"etc."
など	助詞	副助詞	ナド

# NB don't *actually* really understand this, but probably need to see
# it in a bunch of sentences to do so
として	助詞	格助詞	トシテ

月	名詞	一般	ツキ
```

Does that usage make sense?